---
title: 线程中断
date: 2020-06-05 14:56:06
tags:
	- Java
---

## 应用场景 
 1.点击某个桌面应用中的取消按钮时；

   2、某个操作超过了一定的执行时间限制需要终止时；

   3、多个线程做同样的事情，只要一个线程成功其他线程都可以取消时；

  4、一组线程中的一个或多个出现错误导致整组都无法继续时；

  5、当一个应用或服务需要停止时；

<!--more-->

知乎@谢照东

更多场景：

- 比如我们会启动多个线程做同一件事情，抢12306的火车票，我们可能开启多个线程从多个渠道买火车票，只要由一个渠道买到了，我们就会通知取消其他渠道，这个时候需要关闭其他线程

- 很多线程的运行模式是死循环，比如在生产车/消费者模式中，消费者主体就是一个死循环，它不停地从队列中接受任务，执行任务，在停止程序时，我们需要一种“优雅”的方法以关闭该线程
- 在一些场景中，比如从第三方服务器查询一个结果，我们希望在限定的时间内得到结果，如果得不到，我们会希望取消该任务

## 例子
1.在廖大的教程里有这样一个例子：
```
public class InterruptThread {
	public static void main(String[] args) throws InterruptedException {
		Thread thread = new MyNewThread();
		thread.start();
		Thread.sleep(1);//main线程sleep
		thread.interrupt(); //thread线程中断
		thread.join();//等待thread线程结束
		System.out.println("end");
	}
}

class MyNewThread extends Thread{
	@Override
	public void run() {
		// TODO Auto-generated method stub
		int n = 0;
		while(!isInterrupted()) {
			n++;
			System.out.println(n+"hello");
		}
	}
}
```
执行了几次，程序结果都不相同，说明以下问题：
- 控制台1毫秒能打印的行数不是确定的数哦

2. 另一个例子
```
public static void main(String[] args) throws InterruptedException {
		System.out.println("1111");
		Thread.currentThread().interrupt();
		System.out.println("2222");
	}
```
结果：
```
1111
2222
```
说明单纯调用interrupt方法，线程不会被中断

~~## 2.正确中断线程的方法~~
~~### 2.1 使用共享变量~~
~~在这种方式中，之所以引入共享变量，是因为该变量可以被多个执行相同任务的线程用来作为是否中断的信号，通知中断线程的执行。~~

[https://zhuanlan.zhihu.com/p/27857336](https://zhuanlan.zhihu.com/p/27857336)

## 2.线程对interrupt()的反应：
### RUNNABLE状态
如果线程在运行中，interrupt()只是会设置线程的中断标志位，没有任何其他作用。
线程应该在运行过程中合适的位置检查中断标志位，比如说，如果主体代码是一个循环，可以在循环开始处进行检查，如下所示：
```
public class InterruptRunnableDemo extends Thread {
    @Override
    public void run() {
        while (!Thread.currentThread().isInterrupted()) {
            // ... 单次循环代码
        }
        System.out.println("done ");
    }
    public static void main(String[] args) throws InterruptedException {
        Thread thread = new InterruptRunnableDemo();
        thread.start();
        Thread.sleep(1000);
        thread.interrupt();
    }
}
```
### WAITING/TIMED_WAITING
这两个状态时，对线程对象调用interrupt()会使得该线程抛出InterruptedException
>P.s 这就是sleep等方法必须 用try catch包围的原因？

，需要注意的是，抛出异常后，中断标志位会被清空(线程的中断标志位会由true重置为false，因为线程为了处理异常已经重新处于就绪状态。)，而不是被设置。比如说，执行如下代码：
```
Thread t = new Thread (){
    @Override
    public void run() {
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
        //exception被捕获，但是为输出为false 因为标志位会被清空
            System.out.println(isInterrupted());
        }
    }        
};
t.start();
try {
    Thread.sleep(100);
} catch (InterruptedException e) {
}
t.interrupt();//置为true
```

捕获到InterruptedException，通常表示希望结束该线程，线程大概有两种处理方式：

1.向上传递该异常，这使得该方法也变成了一个可中断的方法，需要调用者进行处理
2.有些情况，不能向上传递异常，比如Thread的run方法，它的声明是固定的，不能抛出任何受检异常，这时，应该捕获异常，进行合适的清理操作，清理后，一般应该调用Thread的interrupt方法设置中断标志位，使得其他代码有办法知道它发生了中断

第一种方法：
```
//抛出中断异常，由调用者捕获
public void interruptibleMethod() throws InterruptedException{
    // ... 包含wait, join 或 sleep 方法
    Thread.sleep(1000);
}
```
第二种方法:
```
public class InterruptWaitingDemo extends Thread {
    @Override
    public void run() {
        while (!Thread.currentThread().isInterrupted()) {
            try {
                // 模拟任务代码
                Thread.sleep(2000);
            } catch (InterruptedException e) {
                // ... 清理操作
                System.out.println(isInterrupted());//false
                // 重设中断标志位为true
                Thread.currentThread().interrupt();
            }
        }
        System.out.println(isInterrupted());//true
    }

    public static void main(String[] args) {
        InterruptWaitingDemo thread = new InterruptWaitingDemo();
        thread.start();
        try {
            Thread.sleep(100);
        } catch (InterruptedException e) {
        }
        thread.interrupt();
    }
}
```
![](https://upload-images.jianshu.io/upload_images/19092361-6a41858f4b0f06a3.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
如果不加`Thread.currentThread().interrupt();`这一句，线程一直不停止。

![](https://upload-images.jianshu.io/upload_images/19092361-e368e99f01f61e02.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
死循环了？


看Debug信息：
![](https://upload-images.jianshu.io/upload_images/19092361-fa842b40faf695fb.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
sleep时抛出中断异常。
为什么Object.wait、Thread.sleep和Thread.join都会抛出InterruptedException?首先，这个异常的意思是表示一个阻塞被其他线程中断了。然后，由于线程调用了interrupt()中断方法，那么Object.wait、Thread.sleep等被阻塞的线程被唤醒以后会通过is_interrupted方法判断中断标识的状态变化，如果发现中断标识为true，则先清除中断标识，然后抛出InterruptedException
**需要注意的是，InterruptedException异常的抛出并不意味着线程必须终止，而是提醒当前线程有中断的操作发生**，至于接下来怎么处理取决于线程本身，比如

1.直接捕获异常不做任何处理
2.将异常往外抛出
3.停止当前线程，并打印异常信息

**懂了：
第一次sleep(2000)，被main()线程中断，此时thread的isInterrupted标志位设置为true，sleep检测到，将isInterrupted标志位设置为false之后，抛出异常，被catch捕获。
那么现在thread的isInterrupted标志位为false，while判断条件为真，不断执行sleep(2000)。**

**所以，可能会出现阻塞状态的线程，正确的中断方式是：**
- **run方法中首先查询是否被中断**
- **用try-catch块捕获可能出现的异常**
- **如果被阻塞，相关处理，然后Thread.currentThread().interrupt();重新设置中断标志位为true**

### BLOCKED
如果线程在等待锁，对线程对象调用interrupt()只是会设置线程的中断标志位，线程依然会处于BLOCKED状态，也就是说，interrupt()并不能使一个在等待锁的线程真正”中断”。我们看段代码：
```
public class InterruptSynchronizedDemo {
    private static Object lock = new Object();//monitor
    private static class A extends Thread {
        @Override
        public void run() {
        	//等待lock锁
            synchronized (lock) {
            		//等待标志位被置为true
                while (!Thread.currentThread().isInterrupted()) {
                }
            }
            System.out.println("exit");
        }
    }
    public static void test() throws InterruptedException {
        synchronized (lock) {//获取锁
        	A a = new A();
            a.start();
            Thread.sleep(1000);
            a.interrupt();
            a.join();
        }
    }
    public static void main(String[] args) throws InterruptedException {
        test();
    }
}
```
为了验证a线程的线程状态没改变，改写了class A的interrupt方法：
```
@Override
        public void interrupt() {
        	// TODO Auto-generated method stub
        	super.interrupt();
        	System.out.println(this.getState());
        }
```
输出：
```
Run
BLOCKED
//并且这个线程一直没有结束Orz
```
>我理解了。。这是个死锁场景啊
>主线程获得锁，然后a线程开启，主线程要a线程结束才会释放锁，但是a线程拥有 锁才会开始，
>┓( ´∀` )┏

修改代码：
```
A a = new A();
a.start();
Thread.sleep(1000);
a.interrupt();
```
主线程不需要等待a线程，那么主线程结束，释放锁，a线程获得锁。
但是！！！！在a线程等锁期间，main线程还是修改了a线程的isInterruped状态！！
因此
```
synchronized (lock) {
  //等待标志位被置为true
     System.out.println("获得锁");
     System.out.println(Thread.currentThread().isInterrupted());
      while (!Thread.currentThread().isInterrupted()) {
          System.out.println("没有被中断");
        		}
     }
    System.out.println("exit");
```
输出结果就是：
```
获得锁
true
exit
```
所以要注意这一点，如果有检查isInterrupted标志位的代码，要注意此线程的标志位。
### NEW/TERMINATE
如果线程尚未启动(NEW)，或者已经结束(TERMINATED)，则调用interrupt()对它没有任何效果，中断标志位也不会被设置。比如说，以下代码的输出都是false。

## 3.Interrupt 原理分析
interrupt代码：
```
public void interrupt() {
        if (this != Thread.currentThread())
            checkAccess();

        synchronized (blockerLock) {
            Interruptible b = blocker;
            if (b != null) {
                interrupt0();           // Just to set the interrupt flag
                b.interrupt(this);
                return;
            }
        }
        interrupt0();
    }
```
这个方法里面，调用了interrupt0()，这个方法在前面分析start方法的时候见过，是一个native方法。
代码如下：
```
JVM_ENTRY(void, JVM_Interrupt(JNIEnv* env, jobject jthread))
  JVMWrapper("JVM_Interrupt");
  // Ensure that the C++ Thread and OSThread structures aren't freed before we operate
  oop java_thread = JNIHandles::resolve_non_null(jthread);
  MutexLockerEx ml(thread->threadObj() == java_thread ? NULL : Threads_lock);
  // We need to re-resolve the java_thread, since a GC might have happened during the
  // acquire of the lock
  JavaThread* thr = java_lang_Thread::thread(JNIHandles::resolve_non_null(jthread));
  if (thr != NULL) {
    Thread::interrupt(thr);
  }
JVM_END
```
这个方法直接调用了Thread::interrupt(thr)，代码如下：
```
void Thread::interrupt(Thread* thread) {
  trace("interrupt", thread);
  debug_only(check_for_dangling_thread_pointer(thread);)
  os::interrupt(thread);
}
```
Thread::interrupt方法调用了os::interrupt方法，这个是调用平台的interrupt方法，这个方法的实现是在 os_*.cpp文件中，其中星号代表的是不同平台，因为jvm是跨平台的，所以对于不同的操作平台，线程的调度方式都是不一样的。我们以os_linux.cpp文件为例。
```
void os::interrupt(Thread* thread) {
  assert(Thread::current() == thread || Threads_lock->owned_by_self(),
    "possibility of dangling Thread pointer");
  //获取本地线程对象
  OSThread* osthread = thread->osthread();
  if (!osthread->interrupted()) {//判断本地线程对象是否为中断
    osthread->set_interrupted(true);//设置中断状态为true
    // More than one thread can get here with the same value of osthread,
    // resulting in multiple notifications.  We do, however, want the store
    // to interrupted() to be visible to other threads before we execute unpark().
    //这里是内存屏障，这块在后续的文章中会剖析；内存屏障的目的是使得interrupted状态对其他线程立即可见
    OrderAccess::fence();
    //_SleepEvent相当于Thread.sleep，表示如果线程调用了sleep方法，则通过unpark唤醒
    ParkEvent * const slp = thread->_SleepEvent ;
    if (slp != NULL) slp->unpark() ;
  }
  // For JSR166. Unpark even if interrupt status already was set
  if (thread->is_Java_thread())
    ((JavaThread*)thread)->parker()->unpark();
  //_ParkEvent用于synchronized同步块和Object.wait()，这里相当于也是通过unpark进行唤醒
  ParkEvent * ev = thread->_ParkEvent ;
  if (ev != NULL) ev->unpark() ;
}
```
内存屏障。。。《并发编程的艺术》开头就说了这个。

通过上面的代码分析可以知道，thread.interrupt()方法实际就是设置一个interrupted状态标识为true、并且通过ParkEvent的unpark方法来唤醒线程。

1.对于synchronized阻塞的线程，被唤醒以后会继续尝试获取锁，如果失败仍然可能被park
2.在调用ParkEvent的park方法之前，会先判断线程的中断状态，如果为true，会清除当前线程的中断标识
3.Object.wait、Thread.sleep、Thread.join会抛出InterruptedException